networkx/algorithms/dag.py

```diff
@@ -355,6 +387,108 @@ def transitive_closure(G):
+    a graph G+ = (V,E+) such that for all directed paths from u to v in G there
+    is a directed path from u to v in G+ and if there is no graph with fewer
+    edges than G+ that satisfies the first condition. When the graph is acyclic
+    this is also the unique minimal graph. The transitive reduction of a cyclic
+    graph may not be unique.
+
+    Parameters
+    ----------
+    G : NetworkX DiGraph
+        Graph
+
+    Returns
+    -------
+    TR : NetworkX DiGraph
+        Graph
+
+    Raises
+    ------
+    NetworkXNotImplemented
+        If G is not directed
+
+    Examples
+    ---------
+
+    >>> G = nx.DiGraph([(0, 1), (0, 2), (0, 4),
+    ...                 (0, 3), (1, 3), (2, 4),
+    ...                 (3, 4), (2, 4), (2, 3)])
+    >>> TR = nx.algorithms.dag.transitive_reduction(G)
+    >>> list(TR.edges())
+    [(0, 1), (0, 2), (1, 3), (2, 3), (3, 4)]
+
+    References
+    ----------
+    .. [1] http://dept-info.labri.fr/~thibault/tmp/0201008.pdf
+    .. [2] https://en.wikipedia.org/wiki/Transitive_reduction#Computing_the_reduction_using_the_closure
+    """
+    def invert_map(mapping):
+        """ point each value to a list of its corresponding keys """
+        from collections import defaultdict
+        pair_list = [(val, key) for key, val in mapping.items()]
+        # Initialize dict of lists
+        groupid2_items = defaultdict(list)
+        # Insert each item into the correct group
+        for groupid, item in pair_list:
+            groupid2_items[groupid].append(item)
+        return groupid2_items
+
+    has_cycles = not nx.is_directed_acyclic_graph(G)
+    if has_cycles:
+        # To handle cycles convert to graph of SCCs, find transitive reduction
+        # and then make a directed cycle for every SCC.
+        G_scc = nx.condensation(G)
+    else:
+        G_scc = G
+
+    # Adapted from
+    # http://stackoverflow.com/questions/17078696/transitive-reduction-python
+    # For each node u, perform DFS consider its set of (non-self) children C.
+    # For each descendant v, of a node in C, remove any edge from u to v.
+    TR_scc = G_scc.copy()
+
+    for parent in TR_scc.nodes():
+        # Remove self loops
+        if TR_scc.has_edge(parent, parent):
+            TR_scc.remove_edge(parent, parent)
+        # For each child of the parent
+        for child in list(TR_scc.successors(parent)):
+            # Preorder nodes includes its argument (no added complexity)
+            for gchild in list(TR_scc.successors(child)):
+                # Remove all edges from parent to non-child descendants
+                for descendant in list(nx.dfs_preorder_nodes(TR_scc, gchild)):
+                    if TR_scc.has_edge(parent, descendant):
+                        TR_scc.remove_edge(parent, descendant)
+
+    if has_cycles:
+        # Uncondense graph.
+        mapping = G_scc.graph['mapping']
+        inv_mapping = invert_map(mapping)
```

Comment:
Might you be able to use the `networkx.utils.groups()` function that I introduced recently?